Full Site Editing: allow toggling of the Site Editor

mattwiebe opened the pull request:
#### Changes proposed in this Pull Request

Adds a toggle in `wp-admin` that allows enabling of the Site Editor and other related functionality tied to `gutenberg_is_fse_theme`.

#### Testing instructions

1. Apply D64218-code to your sandbox
2. Activate `pub/tt1-blocks` on your sandbox.
3. Proxy should be active. (The toggle is only shown to proxied requests)

_NOTE:_ These screenshots are taken in wp-admin. The `(beta)` badge will look different (yellow, with outline) as in #54681

| State  |  Screenshot |
|---|---|
|  Before | <img width="273" alt="Screen Shot 2021-07-20 at 16 07 38" src="https://user-images.githubusercontent.com/195089/126395589-27633776-2e23-4851-b552-241facd438ee.png"> |
| After, enabled  | <img width="274" alt="Screen Shot 2021-07-20 at 16 05 09" src="https://user-images.githubusercontent.com/195089/126395341-46584149-71da-4ce1-9161-91a977d96a20.png"> |
| After, disabled | <img width="274" alt="Screen Shot 2021-07-20 at 16 05 55" src="https://user-images.githubusercontent.com/195089/126395388-563f7c0d-3642-4d73-aec2-d3a04599f05c.png"> |

Note that we'll have to merge Automattic/themes#4340 to revert Blockbase's own Site Editor hiding before making this live.

mattwiebe committed:
first pass

mattwiebe committed:
make CTA in banner a button

mattwiebe committed:
don't show options to non-FSE themes

mattwiebe committed:
make REST requests work

mattwiebe committed:
hide template menu items in endpoint

The REST endpoint needs some help to ensure the template (part) menu items are properly hidden when FSE is disabled.

scruffian commented:
Do we also need to make changes to allow universal themes to be shown in the showcase?

mattwiebe commented:
> Do we also need to make changes to allow universal themes to be shown in the showcase? 

Yes, we would need to make a change there, too. Probably just to hide `tt1-blocks` from non-FSE sites since it's the only non-Universal FSE theme.

mattwiebe committed:
badgify the beta in the submenu item

david-szabo97 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-universal-themes/index.php:
```diff
@@ -0,0 +1,292 @@
+ *
+ * @return void
+ */
+function theme_nag() {
```
Where am I supposed to see this notice? It's not popping up on the themes:

![image](https://user-images.githubusercontent.com/2256104/126649237-4a213880-e390-4117-ac72-fb56fc256289.png)

mattwiebe replied:
Ah! It only show up at `wp-admin/themes.php`, in wp-admin. Once we're ready to allow all users to opt in, we can also put the notice in Calypso.

mattwiebe replied:
Although, we might handle the notice differently here, perhaps prompting someone instead while they're activating an FSE theme.

david-szabo97 reviewed (commented):
Works well. I didn't find the notice though.

kwight commented:
We've now got the FSE Theme Showcase tab from https://github.com/Automattic/wp-calypso/pull/54698 launched but hidden behind the `gutenboarding/site-editor` flag – just something to keep in mind when handling the Calypso side. 👍

mattwiebe commented:
Thanks for the heads-up @kwight 👍

david-szabo97 reviewed (approved):
Notice is working too, so I'll leave my ✅  here 😄

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-universal-themes/index.php:
```diff
@@ -0,0 +1,292 @@
+/**
+ * This is the option name for enabling/disabling.
+ */
+define( 'OPTION_NAME', 'is_fse_activated' );
```
Should the name of this constant be more specific than `OPTION_NAME` ?

mattwiebe replied:
Yeah I was probably just too happy to not have to overly prefix everything, given that we're in a suitable namespace here. I'll make it more descriptive.

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-universal-themes/index.php:
```diff
@@ -0,0 +1,292 @@
+ * @return void
+ */
+function deactivate_core_fse() {
+	delete_option( OPTION_NAME );
```
Is it better to delete the option than set it to `false` ?  Maybe having it set to `false` will allow us to have a better idea of which/how many sites have deactivated core FSE after having it enabled?

mattwiebe replied:
Interesting. Ultimately this winds up never being called, I just have it here as a utility, should it be required by code elsewhere. The WP Options API I'm leveraging POSTs everything to `wp-admin/options.php` and, after further investigation, winds up setting the option to `NULL` when the checkbox is disabled. I'll just duplicate that for the sake of consistency, but I think that for tracking purposes we'll need to apply/remove some blog stickers for tracking Simple Sites.

Addison-Stavlo reviewed (commented):
A few notes/Qs:

1.  Is our goal to limit access to the site editor, or disable full site editing entirely?  Currently, we are restricting access to the site editor but FSE functionality still bleeds over into the post editor.  FSE blocks are loaded conditionally on the `__unstableEnableFullSiteEditingBlocks` setting, which is set by the `gutenberg_supports_block_templates()` function.  This is [always true if the theme supports block templates in general](https://github.com/WordPress/gutenberg/blob/6f7cc24e4be5f84649a1c527324f1905d9c05acf/lib/full-site-editing/full-site-editing.php#L22-L23) regardless of the status of `gutenberg_is_fse_theme`.

2.  Should I expect the front end of my site to behave differently when this is turned off?  Currently, it still renders the custom template and custom template parts I had set up in the site editor even after I turn the Site Editor setting off.

Given 1 & 2 above, it seems like this isn't doing much past hiding access to the site editor as FSE functionality still seems to be in place.  Im not sure if this is our goal here?

3.  Does it make sense for this toggle to be present for all FSE themes? or only ones that are specifically set up to be hybrid/universal?

4.  It seems this option is false by default, meaning we will probably need to add a step in site creation to add this option if we expect users to have access to Site Editor / FSE?  Similarly existing sites will lose access to the site editor when this ships (until they toggle the setting), but since they are all internal that doesn't seem like much of an issue.

mattwiebe committed:
more descriptive option name

mattwiebe commented:
Thanks for the thorough and thoughtful review @Addison-Stavlo 🙌 

> Is our goal to limit access to the site editor, or disable full site editing entirely?

It's to limit access to the Site Editor, although it would be ideal to disable FSE entirely. I'll see if I can also disable FSE blocks in the post editor today, thanks for reminding me about that.

> Should I expect the front end of my site to behave differently when this is turned off? Currently, it still renders the custom template and custom template parts I had set up in the site editor even after I turn the Site Editor setting off.

Great question. It's only our experience with FSE enabled prior to this that will land us in that state. A user would have to 1) enable a universal theme, 2) make a custom template (part), and then 3) disable FSE through the toggle. And even there they effectively land up in the same place as a classic theme: some parts of the theme are not editable. But they've gotten there through their own actions. I think that's ok.

> Does it make sense for this toggle to be present for all FSE themes? or only ones that are specifically set up to be hybrid/universal?

We are only planning to have Universal Themes as FSE themes on dotcom. `tt1-blocks` is the only non-universal block theme on dotcom, and we will be deactivating it shortly.

> this option is false by default, meaning we will probably need to add a step in site creation to add this option if we expect users to have access to Site Editor / FSE?

Yes we will set this during site creation for sites that are opting into FSE. The main point of this is to enable users to activate an FSE theme on dotcom without being enrolled in the whole Site Editor experience, but with the option to later switch over and try it if they want to (the toggle will only be available for proxied requests to start).

> existing sites will lose access to the site editor when this ships (until they toggle the setting), but since they are all internal that doesn't seem like much of an issue.

Yup, that was my thinking. I also checked, and could only find one single Atomic site running one of our FSE themes, so it shouldn't be too disruptive on that said, provided we get this out soon. It's not yet possible to transfer over to Atomic with an FSE theme active.

mattwiebe committed:
set to null on deactivation

mattwiebe committed:
prevent FSE blocks from loading with FSE inactive

mattwiebe committed:
slight reorg

mattwiebe commented:
@Addison-Stavlo I've prevented experimental blocks from being loaded in the post editor with 2fd3789c1b5270bacfc28b391f9dea7f3740ea7b and otherwise I think I've addressed all of your feedback.

mattwiebe committed:
namespace our option name with wpcom_ prefix

Addison-Stavlo commented:
Thanks for the explanations, that all makes sense to me!

> otherwise I think I've addressed all of your feedback.

It looks good!  Im having problems testing at the moment...  patching the diff fails on some of the ETK files (might need a rebase there) and when I supplement that with dev build sync from this branch I still see the experimental blocks when turning off the setting. 🤔

david-szabo97 commented:
> I've prevented experimental blocks from being loaded in the post editor

I can confirm experimental blocks aren't available in the post editor if the site editor experience is off. TLDR: It works 👍

Addison-Stavlo reviewed (approved):
I can confirm the changes work as well!  It looks like when I tested yesterday I was looking at a block that used to be part of the experimental FSE group and is no longer 😆 .

mattwiebe commented:
> when I tested yesterday I was looking at a block that used to be part of the experimental FSE group and is no longer

I was confused too! It was working but I didn't think so.
